packages/flutter/lib/src/cupertino/sheet.dart

```diff
@@ -735,31 +784,46 @@ class _CupertinoDownGestureDetectorState<T> extends State<_CupertinoDownGestureD
   }
 }
 
-class _CupertinoDownGestureController<T> {
+class _CupertinoDragGestureController<T> {
   /// Creates a controller for an iOS-style back gesture.
-  _CupertinoDownGestureController({
+  _CupertinoDragGestureController({
     required this.navigator,
-    required this.controller,
+    required this.downController,
     required this.getIsActive,
     required this.getIsCurrent,
   }) {
     navigator.didStartUserGesture();
   }
 
-  final AnimationController controller;
+  final AnimationController downController;
   final NavigatorState navigator;
   final ValueGetter<bool> getIsActive;
   final ValueGetter<bool> getIsCurrent;
 
   /// The drag gesture has changed by [delta]. The total range of the drag
   /// should be 0.0 to 1.0.
-  void dragUpdate(double delta) {
-    controller.value -= delta;
+  void dragUpdate(double delta, AnimationController upController) {
+    if (downController.value == 1.0 && delta < 0) {
+      // Divide by stretchable range (when dragging upward at max extent).
+      upController.value -=
+          delta / (navigator.context.size!.height * (_kTopGapRatio - _kMaxTopGapRatio));
+    } else {
+      // Divide by size of the sheet.
+      downController.value -=
+          delta /
+          (navigator.context.size!.height - (navigator.context.size!.height * _kTopGapRatio));
+    }
   }
 
   /// The drag gesture has ended with a vertical motion of [velocity] as a
   /// fraction of screen height per second.
-  void dragEnd(double velocity) {
+  void dragEnd(double velocity, AnimationController upController) {
+    // If the sheet is in a stretched state (dragged upward beyond max size),
+    // reverse the stretch to return to the normal max height.
+    if (upController.value > 0) {
+      upController.reverse();
+    }
```

Comment:
Theoretically if the `upController` needs to snap back, then the logic below shouldn't need to happen, right? Can we do a `return;` after this or an `else` to cover the rest of this function?

Reply:
fix this commit

https://github.com/flutter/flutter/pull/168547/commits/7822c8357f54371d66778e83d7f70f05aaccfbdb